x-pack/elastic-agent/pkg/core/plugin/service/app.go

```diff
@@ -0,0 +1,330 @@
+	logLevel     string
+	spec         app.Specifier
+	srv          *server.Server
+	srvState     *server.ApplicationState
+	limiter      *tokenbucket.Bucket
+	startContext context.Context
+	tag          app.Taggable
+	state        state.State
+	reporter     state.Reporter
+
+	uid int
+	gid int
+
+	monitor monitoring.Monitor
+
+	processConfig  *process.Config
+	downloadConfig *artifact.Config
+	retryConfig    *retry.Config
+
+	logger *logger.Logger
+
+	credsPort     int
+	credsWG       sync.WaitGroup
+	credsListener net.Listener
+
+	appLock sync.Mutex
+}
+
+// NewApplication creates a new instance of an applications.
+func NewApplication(
+	ctx context.Context,
+	id, appName, pipelineID, logLevel string,
+	credsPort int,
+	spec app.Specifier,
+	srv *server.Server,
+	cfg *config.Config,
+	logger *logger.Logger,
+	reporter state.Reporter,
+	monitor monitoring.Monitor) (*Application, error) {
+
+	s := spec.Spec()
+	uid, gid, err := s.UserGroup()
+	if err != nil {
+		return nil, err
+	}
+
+	b, _ := tokenbucket.NewTokenBucket(ctx, 3, 3, 1*time.Second)
+	return &Application{
+		bgContext:      ctx,
+		id:             id,
+		name:           appName,
+		pipelineID:     pipelineID,
+		logLevel:       logLevel,
+		spec:           spec,
+		srv:            srv,
+		processConfig:  cfg.ProcessConfig,
+		downloadConfig: cfg.DownloadConfig,
+		retryConfig:    cfg.RetryConfig,
+		logger:         logger,
+		limiter:        b,
+		reporter:       reporter,
+		monitor:        monitor,
+		uid:            uid,
+		gid:            gid,
+		credsPort:      credsPort,
+	}, nil
+}
+
+// Monitor returns monitoring handler of this app.
+func (a *Application) Monitor() monitoring.Monitor {
+	return a.monitor
+}
+
+// State returns the application state.
+func (a *Application) State() state.State {
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+	return a.state
+}
+
+// Name returns application name
+func (a *Application) Name() string {
+	return a.name
+}
+
+// SetState sets the status of the application.
+func (a *Application) SetState(status state.Status, msg string) {
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+	a.setState(status, msg)
+}
+
+// Start starts the application with a specified config.
+func (a *Application) Start(ctx context.Context, t app.Taggable, cfg map[string]interface{}) (err error) {
+	defer func() {
+		if err != nil {
+			// inject App metadata
+			err = errors.New(err, errors.M(errors.MetaKeyAppName, a.name), errors.M(errors.MetaKeyAppName, a.id))
+		}
+	}()
+
+	a.appLock.Lock()
+	defer a.appLock.Unlock()
+
+	cfgStr, err := yaml.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+
+	// already started
+	if a.srvState != nil {
+		a.setState(state.Starting, "Starting")
+		a.srvState.SetStatus(proto.StateObserved_STARTING, a.state.Message)
+		a.srvState.UpdateConfig(string(cfgStr))
+	} else {
+		a.setState(state.Starting, "Starting")
+		a.srvState, err = a.srv.Register(a, string(cfgStr))
+		if err != nil {
+			return err
+		}
+	}
+
+	defer func() {
```

Comment:
My brain tries to process all the defer chaining. Should we perhaps have 2 different methods so the named error is only checked once?